kaitaistruct.py

```diff
@@ -14,8 +17,41 @@
 #
 __version__ = '0.9'
 
+class _KaitaiFromCtor(object):
+    """Adds to class methods to construct it from ctor"""
+    @classmethod
+    def from_file(cls, file, map=True):
+        if isinstance(file, str):
+            return cls.from_file(Path(file), map)
+        elif isinstance(file, Path):
+            with file.open("rb") as f:
+                return cls.from_file(f, map)
```

Comment:
We're losing this "try-catch" logic that we had before, which attempted to close file after failure.

I'm by no mean a Python expert, so I'm not sure if that logic made sense in the first place. Can you explan why have you removed it?

Reply:
>We're losing this "try-catch" logic that we had before, which attempted to close file after failure.

A context manager is used. It is a replacement for `try` `except` and is something like RAII: when the block of a context manager is exitted the resource inits header is `__exit__`ed. In fact context managers are present in python for pretty long time, they were introduced in python 2.5 in 2008.

Reply:
I recall that it was tried once, but I had to revert it: https://github.com/kaitai-io/kaitai_struct_python_runtime/commit/7d0b22688fe73541a4ee1908039322d2aa755b3a — have you checked if it works now?

Reply:
No, I don't check the tests since they are not in this repo.

Reply:
In this case you actually cannot use `with`. In `from_file` you normally want the file to stay open so that the returned `KaitaiStream` can use it, and once the `KaitaiStream` is closed, the underlying file is also closed. But if an exception is thrown while creating the `KaitaiStream`, you need to closed the file manually, because then `from_file` doesn't return a `KaitaiStream` object that the user could close. `with` will *always* close the file at the end of the block, regardless of whether an exception is thrown, which is not the desired behavior here, so the manual `try`/`except` solution is actually necessary.

Reply:
You are completely right.